Add structured_function_doc rule

arangato opened the pull request:
This PR  adds basic implementation of the rule requested in https://github.com/realm/SwiftLint/issues/3801.
It covers the `parameters section` method to document function parameters described in [Documentation Comment Syntax](https://github.com/apple/swift/blob/main/docs/DocumentationComments.md).

jpsim commented on Package.swift:
```diff
@@ -19,6 +19,7 @@ let package = Package(
         .package(url: "https://github.com/jpsim/SourceKitten.git", from: "0.31.1"),
         .package(url: "https://github.com/jpsim/Yams.git", from: "4.0.2"),
         .package(url: "https://github.com/scottrhoyt/SwiftyTextTable.git", from: "0.9.0"),
+        .package(url: "https://github.com/apple/swift-markdown.git", .branch("main")),
```
We can't depend on branches here for published packages. Is there a version we can pin to?

jpsim commented:
Do you plan on addressing the CI failures?

arangato commented:
It seems that [swift-markdown](https://github.com/apple/swift-markdown.git) package is problematic for 2 reasons:
- The only release version there has unstable dependency
- It fails linux builds with an unresolved dependency: ``` swift-markdown: error: product dependency 'cmark-gfm' in package 'swift-cmark' not found ```

I intend to replace swift-markdown with a healthier alternative. Suggestions are welcomed, by the way. (it was initially chosen because [swift-docc](https://github.com/apple/swift-docc) uses it)

I have to admit that I'm a complete novice to SPM. If there is a solution to this dependency issues I'd be super glad if somebody pointed it out for me.

arangato commented on SwiftLint.podspec:
```diff
@@ -9,7 +9,7 @@ Pod::Spec.new do |s|
   s.exclude_files             = '**/file.zip'
   s.ios.deployment_target     = '9.0'
-  s.macos.deployment_target   = '10.10'
+  s.macos.deployment_target   = '10.11'
```
This should probably be done in a separate PR.
By the way, package.swift targets 10.12, so shouldn't we use 10.12 in podspec as well ?
@jpsim WDYT?

jpsim replied:
This is a bit confusing, because the SwiftLint podspec just vends the SwiftLint library. SwiftLint doesn't run on iOS, watchOS or tvOS either, yet this podspec claims to support those platforms.

That's done so iOS/nacOS/tvOS/watchOS Xcode projects using CocoaPods can all add SwiftLint as a dependency, not to embed it into their apps but to run as an Xcode build phase.

So the deployment target here doesn't have to match what's in Package.swift and in fact should just go as far back as possible.

arangato replied:
I see. Thanks for detailed explanation. 
Change to 10.11 stems from using [Down](https://github.com/johnxnguyen/Down). It's far enough, isn't it ?

jpsim replied:
Yes.

jpsim commented:
Can this support the inline parameter list style?

```swift
/// Summary
///
/// - parameter a: First.
/// - parameter b: Second.
```

arangato commented:
> Can this support the inline parameter list style?
> 
> ```swift
> /// Summary
> ///
> /// - parameter a: First.
> /// - parameter b: Second.
> ```

I'll give it a go.

arangato committed:
Add support for inline parameter style

arangato committed:
Add descriptive reasons to structured_function_doc violations

arangato committed:
Fix handling lack of space after parameter keyword

arangato committed:
Detail missing and extra parameters in violation reason

arangato committed:
Fix parsing text intermixed with code

arangato commented:
I've added support for inline parameters. The rule looks for either parameters section or inlined ones. An arbitrary intermix of the two, which is allowed by [documentation syntax](https://github.com/apple/swift/blob/main/docs/DocumentationComments.md), the rule doesn't cover.
I bet such style intermixes are unpopular anyway.

arangato commented:
@jpsim could you please take a look ?

arangato committed:
Add structured-function-doc rule

arangato committed:
Replace 'swift-markdown' package with 'down'

jpsim committed:
Fix analyzer rules with Xcode 13.3

Looks like starting with Xcode 13.3 / Swift 5.6, cursor info requests
started canceling in-flight requests, so we need to pass
`key.cancel_on_subsequent_request: false` to bypass that.

Analyzer rules on Swift 5.6 are extremely slow, however. Not really
usable right now.

jpsim committed:
Run analyzer rules one file at a time

jpsim committed:
Add changelog entry

arangato committed:
Add changelog entry for structured_function_doc opt-in rule

arangato committed:
Merge remote-tracking branch 'origin/master' into add-structured-function-doc-rule

arangato commented:
@marcelofabri could you please have a look ?
